woocommerce/google-listings-and-ads
# PMax Assets: Integrate with APIs to load and save assets

### Changes proposed in this Pull Request:

This PR implements parts of 📌 [Assets form and related components](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-misc) and a part of 📌 [Integrating with backend APIs](https://github.com/woocommerce/google-listings-and-ads/issues/1787#fe-integrating-apis) in #1787.

- Load existing asset groups or suggested assets
- When creating a new campaign, update the assets into the empty asset group of the newly created campaign
- When editing a campaign, create an asset group if needed, and update the assets into the existing or newly created asset group
- The Final URL of the non-empty asset group is not allowed to change 

### Screenshots:

#### 📹 Update a campaign without an existing asset group

https://user-images.githubusercontent.com/17420811/217511157-9d795612-96f1-430c-b8aa-7c50270d30b6.mp4

This video includes:

- Create a new asset group first if needed
- Avoid re-creating another new asset group if the subsequent asset group update is failed
- Update the assets for the campaign
- Enter the same editing page to see if the update is in effect

#### 📹 Update a campaign with an existing asset group

https://user-images.githubusercontent.com/17420811/217512174-a4ff1e16-6ce7-4fa9-abaa-c128e2ce3d34.mp4

This video includes:

- The Final URL of the non-empty asset group is not allowed to change
- Except for the Final URL, update all assets into the existing asset group
- Enter the same editing page to see if the update is in effect

#### 📹 Create a campaign with assets

https://user-images.githubusercontent.com/17420811/217514975-345883b1-fdcc-4ba0-a6c5-11b9f4c91551.mp4

This video includes:

- Avoid re-creating another new campaign if the subsequent asset group update is failed
- Update the assets into the empty asset group of the newly created campaign
- Enter the editing page of the same campaign to see if the assets are saved
- Start creating another campaign and loading the suggested assets with the same Final URL from the Google Ads end

### Detailed test instructions:

#### 💡 Additional notes before testing

- Google Ads side will check if the domain in the Final URL matches the claimed domain of the linked Merchant Center account.
   ![image](https://user-images.githubusercontent.com/17420811/217712679-467331b6-e555-4e9d-9a6f-9b0bf726d6e9.png)
   Therefore, it may need to make the local env have the same domain by switching the local env to be accessible publicly or replacing the domain temporarily. The way I used:
   - Replace [this line](https://github.com/woocommerce/google-listings-and-ads/blob/2f448130ce83497e46102a63e28464505f30e146/js/src/components/paid-ads/convertToAssetGroupUpdateBody.js#L86) with:
      ```js
      final_url: values.final_url.replace( 'your local domain', 'GMC domain' ),
      ```
   - Add the following at [this line](https://github.com/woocommerce/google-listings-and-ads/blob/2f448130ce83497e46102a63e28464505f30e146/src/Ads/AssetSuggestionsService.php#L191):
      ```php
      $final_url = str_replace( 'your local domain', 'GMC domain', $final_url );
      ```
   - Add the following at the end of **google-listings-and-ads.php** if got the "cURL error 60" issue with `wp_remote_get`:
      ```php
      add_filter( 'https_ssl_verify', '__return_false' );
      add_filter( 'https_local_ssl_verify', '__return_false' );
      ```
- Please note that the images may not be loaded properly the first few times after saving new asset images to Google Ads end. Looks like my network or Google's CDN takes a little more time to load the images correctly.

#### Test creation page

1. Go to the campaign creation page.
2. Select a Final URL that has never been used for any asset groups before.
3. Fill in all asset fields but enter at least an invalid content to text assets, for example, emoji chars.
4. After clicking "Saving changes", it should result in a submission failure.
5. Remove the invalid chars entered in step 3.
6. Click "Saving changes" again to see if the assets are saved.
7. After back to the Dashboard page, it should be no duplicate campaign even if the submission failure occurred intentionally during steps 3-4.
8. Go to the campaign creation page again.
9. Select the same Final URL as step 2 and load the suggested assets.
10. Check if the suggested assets are the same as the previously created ones.

#### Test editing page

1. Create a new campaign by skipping adding assets.
2. Go to the editing page of the newly created campaign. (Currently, the asset group of this campaign is empty.)
3. Fill in all asset fields.
4. Click "Saving changes" again to see if the assets are saved.
5. Go to Google Ads admin page
   1. Navigate to the assets management tab of the same campaign
   2. Remove all asset groups
      ![XD8vmD.png](https://user-images.githubusercontent.com/2488994/212314911-68a7b90d-6c67-4930-a292-3463a878295b.png)
6. Reload GLA web page and go to the editing page of the same campaign. (Currently, this campaign has no asset groups.)
7. Add at least an invalid content to text assets, for example, emoji chars.
8. After clicking "Saving changes", it should result in a submission failure.
9. Remove the invalid chars entered in step 7.
10. Click "Saving changes" again to see if the assets are saved.
11. Switch back to Google Ads admin page and reload it.
    - There should be only one asset group linked to the same campaign.
    - It should be no other (empty) asset groups campaign even if the submission failure occurred intentionally during steps 7-8.

### Changelog entry
